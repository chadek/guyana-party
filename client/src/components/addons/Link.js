import React from 'react'
import PropTypes from 'prop-types'
import { Link as GatsbyLink } from 'gatsby'

function Link({ children, to, blank, noprefetch, ...other }) {
  const A = blank ? (
    <a
      aria-label={`${to} (sâ€™ouvre dans un nouvel onglet)`}
      href={to}
      rel="noreferrer noopener nofollow"
      target="_blank"
      {...other}
    >
      {children}
    </a>
  ) : (
    <a href={to} {...other}>
      {children}
    </a>
  )

  // Tailor the following test to your environment.
  // This example assumes that any internal link (intended for Gatsby)
  // will start with exactly one slash, and that anything else is external.
  const internal = /^\/(?!\/)/.test(to)

  // Use Gatsby Link for internal links, and <a> for others
  if (internal) {
    const file = /\.[0-9a-z]+$/i.test(to)
    if (file || noprefetch) return A

    return (
      <GatsbyLink to={to} {...other}>
        {children}
      </GatsbyLink>
    )
  }
  return A
}

Link.propTypes = {
  children: PropTypes.node.isRequired,
  to: PropTypes.string.isRequired,
  blank: PropTypes.bool
}

export default Link
