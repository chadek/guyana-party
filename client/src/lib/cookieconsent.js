import 'cookieconsent/build/cookieconsent.min.css'

if (typeof window !== 'undefined') {
  // Import non server side rendering cookieconsent
  import('cookieconsent').then(() => {
    window.cookieconsent.initialise({
      palette: {
        popup: { background: '#fff' },
        button: { background: '#3f51b5' }
      },
      theme: 'edgeless',
      position: 'bottom-left',
      content: {
        message:
          "Nous utilisons des cookies afin d'améliorer votre expérience sur le site et ainsi vous proposer un contenu adapté.",
        link: 'En savoir plus',
        href: '/mentions-legales/#cookies',
        dismiss: 'OK'
      }
    })
  })
}
