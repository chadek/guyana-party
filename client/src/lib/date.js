import { zonedTimeToUtc, utcToZonedTime, format } from 'date-fns-tz'
import fr from 'date-fns/locale/fr'

export const userTZ = () => {
  try {
    let offset = new Date().getTimezoneOffset()
    const o = Math.abs(offset)
    offset =
      (offset < 0 ? '+' : '-') +
      ('00' + Math.floor(o / 60)).slice(-2) +
      ':' +
      ('00' + (o % 60)).slice(-2)
    return tzList.filter(t => t.offset === offset)[0].tzCode
  } catch (error) {
    console.log(error)
    return 'Europe/London'
  }
}

export const toUTCIsoDate = (date, timeZone) => zonedTimeToUtc(date, timeZone).toISOString()

export const toZonedTime = (date, timeZone) => utcToZonedTime(date, timeZone)

export const dateToStr = (date, timeZone, formatStr = 'd MMM yyyy à HH:mm (zzz)') => {
  const locale = fr
  if (timeZone) return format(utcToZonedTime(date, timeZone), formatStr, { locale, timeZone })
  return format(new Date(date), formatStr, { locale })
}

export const formatStart = startDate => {
  const tz = userTZ()
  const d = dateToStr(startDate, tz, 'EEEE d MMMM yyyy').split(' ')
  let day = d[0]
  day = day.charAt(0).toUpperCase() + day.slice(1)
  const dayNb = d[1]
  let month = d[2]
  month = month.charAt(0).toUpperCase() + month.slice(1)
  const year = d[3]
  return `${day} ${dayNb} ${month} ${year}`
}

export const formatPlage = ({ startDate, endDate }, showUTC = false) => {
  const tz = userTZ()
  const start = dateToStr(startDate, tz, 'dd MM HH:mm (zzz)').split(' ')
  const end = dateToStr(endDate, tz, 'dd MM HH:mm').split(' ')
  const utc = showUTC ? ` ${start[3]}` : ''
  if (start[0] + start[1] === end[0] + end[1]) {
    return `Le ${start[0]}/${start[1]} de ${start[2]} à ${end[2]}${utc}`
  }
  return `Du ${start[0]}/${start[1]} à ${start[2]} au ${end[0]}/${end[1]} à ${end[2]}${utc}`
}

export const days = [
  { label: 'Lundi', value: 'mon' },
  { label: 'Mardi', value: 'tue' },
  { label: 'Mercredi', value: 'wed' },
  { label: 'Jeudi', value: 'thu' },
  { label: 'Vendredi', value: 'fri' },
  { label: 'Samedi', value: 'sat' },
  { label: 'Dimanche', value: 'sun' }
]

export const tzList = [
  {
    offset: '-11:00',
    label: '(GMT-11:00) Pago Pago',
    tzCode: 'Pacific/Pago_Pago'
  },
  {
    offset: '-10:00',
    label: '(GMT-10:00) Hawaii Time',
    tzCode: 'Pacific/Honolulu'
  },
  { offset: '-10:00', label: '(GMT-10:00) Tahiti', tzCode: 'Pacific/Tahiti' },
  {
    offset: '-09:00',
    label: '(GMT-09:00) Alaska Time',
    tzCode: 'America/Anchorage'
  },
  {
    offset: '-08:00',
    label: '(GMT-08:00) Pacific Time',
    tzCode: 'America/Los_Angeles'
  },
  {
    offset: '-07:00',
    label: '(GMT-07:00) Mountain Time',
    tzCode: 'America/Denver'
  },
  {
    offset: '-06:00',
    label: '(GMT-06:00) Central Time',
    tzCode: 'America/Chicago'
  },
  {
    offset: '-05:00',
    label: '(GMT-05:00) Eastern Time',
    tzCode: 'America/New_York'
  },
  {
    offset: '-04:00',
    label: '(GMT-04:00) Atlantic Time - Halifax',
    tzCode: 'America/Halifax'
  },
  {
    offset: '-03:00',
    label: '(GMT-03:00) Buenos Aires',
    tzCode: 'America/Argentina/Buenos_Aires'
  },
  {
    offset: '-02:00',
    label: '(GMT-02:00) Noronha',
    tzCode: 'America/Noronha'
  },
  { offset: '-01:00', label: '(GMT-01:00) Azores', tzCode: 'Atlantic/Azores' },
  { offset: '+00:00', label: '(GMT+00:00) London', tzCode: 'Europe/London' },
  { offset: '+01:00', label: '(GMT+01:00) Berlin', tzCode: 'Europe/Berlin' },
  {
    offset: '+02:00',
    label: '(GMT+02:00) Helsinki',
    tzCode: 'Europe/Helsinki'
  },
  {
    offset: '+03:00',
    label: '(GMT+03:00) Istanbul',
    tzCode: 'Europe/Istanbul'
  },
  { offset: '+04:00', label: '(GMT+04:00) Dubai', tzCode: 'Asia/Dubai' },
  { offset: '+04:30', label: '(GMT+04:30) Kabul', tzCode: 'Asia/Kabul' },
  {
    offset: '+05:00',
    label: '(GMT+05:00) Maldives',
    tzCode: 'Indian/Maldives'
  },
  {
    offset: '+05:30',
    label: '(GMT+05:30) India Standard Time',
    tzCode: 'Asia/Calcutta'
  },
  {
    offset: '+05:45',
    label: '(GMT+05:45) Kathmandu',
    tzCode: 'Asia/Kathmandu'
  },
  { offset: '+06:00', label: '(GMT+06:00) Dhaka', tzCode: 'Asia/Dhaka' },
  { offset: '+06:30', label: '(GMT+06:30) Cocos', tzCode: 'Indian/Cocos' },
  { offset: '+07:00', label: '(GMT+07:00) Bangkok', tzCode: 'Asia/Bangkok' },
  {
    offset: '+08:00',
    label: '(GMT+08:00) Hong Kong',
    tzCode: 'Asia/Hong_Kong'
  },
  {
    offset: '+08:30',
    label: '(GMT+08:30) Pyongyang',
    tzCode: 'Asia/Pyongyang'
  },
  { offset: '+09:00', label: '(GMT+09:00) Tokyo', tzCode: 'Asia/Tokyo' },
  {
    offset: '+09:30',
    label: '(GMT+09:30) Central Time - Darwin',
    tzCode: 'Australia/Darwin'
  },
  {
    offset: '+10:00',
    label: '(GMT+10:00) Eastern Time - Brisbane',
    tzCode: 'Australia/Brisbane'
  },
  {
    offset: '+10:30',
    label: '(GMT+10:30) Central Time - Adelaide',
    tzCode: 'Australia/Adelaide'
  },
  {
    offset: '+11:00',
    label: '(GMT+11:00) Eastern Time - Melbourne, Sydney',
    tzCode: 'Australia/Sydney'
  },
  { offset: '+12:00', label: '(GMT+12:00) Nauru', tzCode: 'Pacific/Nauru' },
  {
    offset: '+13:00',
    label: '(GMT+13:00) Auckland',
    tzCode: 'Pacific/Auckland'
  },
  {
    offset: '+14:00',
    label: '(GMT+14:00) Kiritimati',
    tzCode: 'Pacific/Kiritimati'
  }
]
