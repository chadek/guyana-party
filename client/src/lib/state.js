import { State, subscribe } from 'jstates-react'

export const snackState = new State('snackState', { msg: '', variant: '' })

export { subscribe }
