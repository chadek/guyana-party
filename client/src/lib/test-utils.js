import React from 'react'
import PropTypes from 'prop-types'
import { render } from 'enzyme'

import { ThemeProvider } from 'styled-components'
import { theme, Style } from '../components/styles/Style'

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={theme}>
    <>
      <Style />
      {children}
    </>
  </ThemeProvider>
)

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
// export * from 'enzyme'
export { shallow } from 'enzyme'
export { expect } from 'chai'

// override render method
export { customRender as render }
