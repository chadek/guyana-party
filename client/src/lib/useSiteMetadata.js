import { useStaticQuery, graphql } from 'gatsby'

export default () => {
  const { site } = useStaticQuery(
    graphql`
      query SiteMetaData {
        site {
          siteMetadata {
            title
            titleTemplate
            description
            author
            keywords
          }
        }
      }
    `
  )
  return site.siteMetadata
}
