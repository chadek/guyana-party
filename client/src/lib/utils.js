import axios from 'axios'
import Cookies from 'js-cookie'
import dompurify from 'dompurify'
import Showdown from 'showdown'
import md5 from 'md5'
import cogoToast from 'cogo-toast'

export const MISSING_TOKEN_ERR = 'Token de connexion requis'

export const reload = () => {
  if (typeof window !== 'undefined') window.location.reload()
}

export const gravatar = email => `https://www.gravatar.com/avatar/${md5(email)}?d=retro`

export const purify = async dirty => dompurify.sanitize(dirty)

export const markToSafeHTML = markdown => {
  const converter = new Showdown.Converter({
    tables: true,
    simplifiedAutoLink: true,
    strikethrough: true,
    tasklists: true
  })
  return purify(converter.makeHtml(markdown))
}

export const getUID = () => Cookies.get('gp_uid')

export const getToken = () => ({ jwt: Cookies.get('gp_jwt'), uid: getUID() })

export const fetcher = url => axios.get(url).then(r => r.data)

export const axiosGet = (url, next, fallback) =>
  axios
    .get(url)
    .then(payload => next(payload))
    .catch(fallback)

export const axiosPost = ({ url, data }, next, fallback) => {
  const jwt = Cookies.get('gp_jwt')
  if (!jwt) return fallback(MISSING_TOKEN_ERR)

  const options = { headers: { authorization: `bearer ${jwt}` } }

  return axios.post(url, data, options).then(next).catch(fallback)
}

export const axiosPut = ({ url, data }, next, fallback) => {
  const jwt = Cookies.get('gp_jwt')
  if (!jwt) return fallback(MISSING_TOKEN_ERR)

  const options = { headers: { authorization: `bearer ${jwt}` } }

  return axios.put(url, data, options).then(next).catch(fallback)
}

export const axiosDelete = (url, next, fallback) => {
  const jwt = Cookies.get('gp_jwt')
  if (!jwt) return fallback(MISSING_TOKEN_ERR)

  const options = { headers: { authorization: `bearer ${jwt}` } }

  return axios.delete(url, options).then(next).catch(fallback)
}

/* Convert the coordinate (lat or lng) to DMS (degrees minutes seconds). */
const toDMS = coord => {
  const absolute = Math.abs(coord)
  const degrees = Math.floor(absolute)
  const minutesNotTruncated = (absolute - degrees) * 60
  const minutes = Math.floor(minutesNotTruncated)
  const seconds = Math.floor((minutesNotTruncated - minutes) * 60)
  return `${degrees}° ${minutes}′ ${seconds}″`
}

export const gpsCoords = (lat, lng) => {
  const northSouth = `${toDMS(lat)} ${Math.sign(lat) >= 0 ? 'N' : 'S'}`
  const eastWest = `${toDMS(lng)} ${Math.sign(lng) >= 0 ? 'E' : 'W'}`
  return `${northSouth}, ${eastWest}`
}

export const compress = (files, next) => {
  if (typeof window === 'undefined') return // window not defined in ssr
  const Compress = require('client-compress') // eslint-disable-line global-require
  const compressor = new Compress({ targetSize: 1.0, quality: 0.75 })
  compressor.compress(files).then(data => next(data))
}

export const scrollTo = selector => {
  const target = document.querySelector(selector)
  if (target) target.scrollIntoView({ behavior: 'smooth', block: 'end' })
}

export const toast = (msg, variant = 'info', heading = '') => {
  const { hide } = cogoToast[variant](msg, {
    hideAfter: 5,
    heading: heading || (variant === 'error' ? 'Erreur !' : ''),
    onClick() {
      hide()
    }
  })
}
