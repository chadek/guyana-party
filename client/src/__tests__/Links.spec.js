import React from 'react'
import Link from '../components/addons/Link'
import { expect, shallow } from '../lib/test-utils'

describe('<Links/>', () => {
  it('should render "internal" links correctly', () => {
    const wrapper = shallow(<Link to="/">My Link</Link>)
    expect(wrapper).to.matchSnapshot()
  })

  it('should render "internal file" links correctly', () => {
    const wrapper = shallow(<Link to="/test.pdf">My Link</Link>)
    expect(wrapper).to.matchSnapshot()
  })

  it('should render "internal links with noprefetch" correctly', () => {
    const wrapper = shallow(
      <Link noprefetch to="/lieu/id">
        My Link
      </Link>
    )
    expect(wrapper).to.matchSnapshot()
  })

  it('should render "external" links correctly', () => {
    const wrapper = shallow(
      <Link blank to="https://www.google.com">
        My Link
      </Link>
    )
    expect(wrapper).to.matchSnapshot()
  })

  it('should render links with custom props', () => {
    const wrapper = shallow(
      <Link prop1="other" prop2="prop2" prop3={() => {}} to="/">
        My Link
      </Link>
    )
    expect(wrapper).to.matchSnapshot()
  })
})
