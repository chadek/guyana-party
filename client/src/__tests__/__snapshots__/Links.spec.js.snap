// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Links/> should render "external" links correctly 1`] = `
<a
  aria-label="https://www.google.com (sâ€™ouvre dans un nouvel onglet)"
  href="https://www.google.com"
  rel="noreferrer noopener nofollow"
  target="_blank"
>
  My Link
</a>
`;

exports[`<Links/> should render "internal file" links correctly 1`] = `
<a
  href="/test.pdf"
>
  My Link
</a>
`;

exports[`<Links/> should render "internal links with noprefetch" correctly 1`] = `
<a
  href="/lieu/id"
>
  My Link
</a>
`;

exports[`<Links/> should render "internal" links correctly 1`] = `
<mockConstructor
  to="/"
>
  My Link
</mockConstructor>
`;

exports[`<Links/> should render links with custom props 1`] = `
<mockConstructor
  prop1="other"
  prop2="prop2"
  prop3={[Function]}
  to="/"
>
  My Link
</mockConstructor>
`;
